src/utils/builder.ts

```diff
@@ -69,46 +69,131 @@ export function setCriteriaElm(pathway: Pathway, elm: object): void {
   pathway.elm.criteria = elm;
 }
 
-function addState(pathway: Pathway, key?: string): string {
+export function createState(key?: string): State {
   if (!key) key = shortid.generate();
   const state: State = {
-    key: key,
-    label: '',
+    key,
+    label: 'New Node',
     transitions: []
   };
-  pathway.states[key] = state;
 
-  return key;
+  return state;
 }
 
-export function addBranchState(pathway: Pathway): string {
-  const key = shortid.generate();
-  return addState(pathway, key);
+export function addState(pathway: Pathway, state: State): Pathway {
+  return {
+    ...pathway,
+    states: {
+      ...pathway.states,
+      [state.key as string]: state
+    }
+  };
 }
 
-export function addGuidanceState(pathway: Pathway): string {
-  const key = shortid.generate();
-  addState(pathway, key);
-  makeBranchStateGuidance(pathway, key);
-  return key;
+export function addGuidanceState(pathway: Pathway): Pathway {
+  const state = createState();
+  const newPathway = addState(pathway, state);
+
+  return makeBranchStateGuidance(newPathway, state.key as string);
 }
 
-export function setStateLabel(pathway: Pathway, key: string, label: string): void {
-  pathway.states[key].label = label;
+export function setStateLabel(pathway: Pathway, key: string, label: string): Pathway {
```

Comment:
Is there a reason this changed?

Reply:
Yes. The previous implementation was modifying the pathway in place, but React doesn't know how to handle that. React only knows to re-render when it sees a new object. That's why in Redux we create new objects in actions, and that's what we have to do here.

Reply:
That's really frustrating...